coreneuron/io/lfp.hpp

```diff
@@ -0,0 +1,217 @@
+#ifndef LFP_H
+#define LFP_H
+
+#include <cmath>
+#include <array>
+#include <iostream>
+#include <vector>
+
+#include "coreneuron/nrnconf.h"
+#include "coreneuron/utils/nrn_assert.h"
+#include "coreneuron/mpi/nrnmpidec.h"
+
+namespace coreneuron {
+
+using F = double;
+using Point3D = std::array<F, 3>;
+using Point3Ds = std::vector<Point3D>;
+
+F dot(const Point3D& p1, const Point3D& p2) {
+    return p1[0] * p2[0] + p1[1] * p2[1] + p1[2] * p2[2];
+}
+
+F norm(const Point3D& p1) {
+    return std::sqrt(dot(p1, p1));
+}
+
+Point3D barycenter(const Point3D& p1, const Point3D& p2) {
+    return {0.5 * (p1[0] + p2[0]), 0.5 * (p1[1] + p2[1]), 0.5 * (p1[2] + p2[2])};
+}
+
+Point3D paxpy(const Point3D& p1, const F alpha, const Point3D& p2) {
+    return {p1[0] + alpha * p2[0], p1[1] + alpha * p2[1], p1[2] + alpha * p2[2]};
+}
+
+/**
+ *
+ * \tparam Point3D Type of 3D point
+ * \tparam F Floating point type
+ * \param e_pos electrode position
+ * \param seg_pos segment position
+ * \param radius segment radius
+ * \param f conductivity factor 1/([4 pi] * [conductivity])
+ * \return Resistance of the medium from the segment to the electrode.
+ */
+F point_source_lfp_factor(const Point3D& e_pos, const Point3D& seg_pos, const F radius, const F f) {
+    nrn_assert(radius >= 0.0);
+    Point3D es = paxpy(e_pos, -1.0, seg_pos);
+    return f / std::max(norm(es), radius);
+}
+
+/**
+ *
+ * \tparam Point3D Type of 3D point
+ * \tparam F Floating point type
+ * \param e_pos electrode position
+ * \param seg_pos segment position
+ * \param radius segment radius
+ * \param f conductivity factor 1/([4 pi] * [conductivity])
+ * \return Resistance of the medium from the segment to the electrode.
+ */
+F line_source_lfp_factor(const Point3D& e_pos,
+                         const Point3D& seg_0,
+                         const Point3D& seg_1,
+                         const F radius,
+                         const F f) {
+    nrn_assert(radius >= 0.0);
+    Point3D dx = paxpy(seg_1, -1.0, seg_0);
+    Point3D de = paxpy(e_pos, -1.0, seg_0);
+    F dx2(dot(dx, dx));
+    F dxn(std::sqrt(dx2));
+    if (dxn < std::numeric_limits<F>::epsilon()) {
+        return point_source_lfp_factor(e_pos, seg_0, radius, f);
+    }
+    F de2(dot(de, de));
+    F mu(dot(dx, de) / dx2);
+    Point3D de_star(paxpy(de, -mu, dx));
+    F de_star2(dot(de_star, de_star));
+    F q2(de_star2 / dx2);
+
+    F delta(mu * mu - (de2 - radius * radius) / dx2);
+    F one_m_mu(1.0 - mu);
+    auto log_integral = [&q2, &dxn](F a, F b) {
+        if (q2 < std::numeric_limits<F>::epsilon()) {
+            if (a * b <= 0) {
+                throw std::invalid_argument("Log integral: invalid arguments " + std::to_string(b) +
```

Comment:
You should use `std::ostringstream` instead of using string concatenation.

Reply:
Done.